python/ray/autoscaler/aws/config.py

```diff
@@ -332,15 +338,97 @@ def _get_vpc_id_or_die(config, subnet_id):
-            "Values": [vpc_id]
+            "Values": unique_vpc_ids
         }]))
-    for sg in existing_groups:
-        if sg.group_name == group_name:
-            return sg
+    filtered_groups = [
+        sg for sg in existing_groups if sg.group_name in unique_group_names
+    ]
+    return filtered_groups
+
+
+def _create_security_group(config, vpc_id, group_name):
+    client = _client("ec2", config)
+    client.create_security_group(
+        Description="Auto-created security group for Ray workers",
+        GroupName=group_name,
+        VpcId=vpc_id)
+    security_group = _get_security_group(config, vpc_id, group_name)
+    # only create tags AFTER getting the security group to avoid occasional
+    # eventual consistency failures that occur when tagging immediately after
+    # security group creation (claiming the security group does not exist)
+    _tag_security_group(config, security_group.id)
+    logger.info("_create_security_group: Created new security group {} ({})"
+                .format(security_group.group_name, security_group.id))
+    assert security_group, "Failed to create security group"
+    return security_group
+
+
+def _tag_security_group(config, security_group_id):
+    client = _client("ec2", config)
+    client.create_tags(
+        Resources=[security_group_id],
+        Tags=[{
+            "Key": TAG_RAY_CLUSTER_NAME,
+            "Value": config["cluster_name"]
+        }])
+
+
+def _upsert_security_group_rules(conf, security_groups):
+    sgids = {sg.id for sg in security_groups.values()}
+    # sort security group items for deterministic inbound rule config order
+    # (mainly supports more precise stub-based boto3 unit testing)
+    for node_type, sg in sorted(security_groups.items()):
+        sg = security_groups[node_type]
+        if not sg.ip_permissions:
+            _update_inbound_rules(sg, sgids, conf)
+
+
+def _update_inbound_rules(target_security_group, sgids, config):
+    extended_rules = config["provider"] \
+        .get("security_group", {}) \
+        .get("IpPermissions", [])
+    ip_permissions = _create_default_inbound_rules(sgids, extended_rules)
+    target_security_group.authorize_ingress(IpPermissions=ip_permissions)
+
+
+def _create_default_inbound_rules(trusted_sgids, extended_rules=[]):
+    ip_permissions = [{
+        "FromPort": -1,
+        "ToPort": -1,
+        "IpProtocol": "-1",
+        "UserIdGroupPairs": [
+            {
+                "GroupId": security_group_id
+            } for security_group_id in sorted(trusted_sgids)
+            # sort security group IDs for deterministic IpPermission models
+            # (mainly supports more precise stub-based boto3 unit testing)
+        ]
+    }]
+
+    ip_permissions.extend([{
+        "FromPort": 22,
+        "ToPort": 22,
+        "IpProtocol": "tcp",
+        "IpRanges": [{
+            "CidrIp": "0.0.0.0/0"
+        }]
+    }])
+
+    ip_permissions.extend(extended_rules)
+
+    return ip_permissions
```

Comment:
Ooc, would it make sense to put the ssh rule into the original `ip_permissions` definition?

Alternate thought: maybe we name all of them so it is obvious what is what and then flatten at the end?
```suggestion
    intracluster_rules = [{
        "FromPort": -1,
        "ToPort": -1,
        "IpProtocol": "-1",
        "UserIdGroupPairs": [
            {
                "GroupId": security_group_id
            } for security_group_id in sorted(trusted_sgids)
            # sort security group IDs for deterministic IpPermission models
            # (mainly supports more precise stub-based boto3 unit testing)
        ]
    }]

    ssh_rules = [{
        "FromPort": 22,
        "ToPort": 22,
        "IpProtocol": "tcp",
        "IpRanges": [{
            "CidrIp": "0.0.0.0/0"
        }]
    }])

    merged_rules = itertools.chain(
        intracluster_permissions,
        ssh_permissions,
        extended_rules,
    )

    return list(merged_rules)
```

Reply:
I think I like the named version of it a bit more, and may take that one step further to wrap small generation methods around each named section.  Ultimately I'd like the relationship to be clear here that we basically have a 3-step process to:

1. Generate inbound rules with trusted security groups. 
2. Allow all inbound connections to a required default set of protocols (currently only SSH).
3. Apply any custom inbound rules specified in user config.